Shader viewer buttons fixes

w-pearson opened the pull request:
<!--
Before submitting a pull request you are strongly recommended to read the
docs/CONTRIBUTING.md file which gives some information on how to prepare a
change:

https://github.com/baldurk/renderdoc/blob/v1.x/docs/CONTRIBUTING.md

For small changes you don't have to read the document end to end, but should at
least look at the sections on how to ensure your code and commits are formatted
according to the style requirements.
-->

## Description

<!--
Describe here what your pull request changes and why it should happen. For small
changes which are obvious this can just be a line or two - even the commit
message is sometimes enough.
-->
This PR consists of three small improvements to the shader viewer/debugger and its buttons.

In the second commit, I changed the popup mode for the execute forwards/backwards buttons. The old mode has an arrow that's used to access the menu while the button itself performs the default action. However, the execute forwards/backwards buttons have no default action, so clicking them does nothing, meaning the tiny arrow needs to be clicked instead. That's highly inconvenient. Now, clicking the button always opens the menu. (Theoretically, a default action could be set, but in my opinion always opening the menu is better.)

<table role="presentation">
<tr>
<th align="right">Before</td>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231314002-d2aba119-27a1-4e92-aa66-248eed542a57.png"><img src="https://user-images.githubusercontent.com/127789813/231314002-d2aba119-27a1-4e92-aa66-248eed542a57.png" alt="image" style="max-width: 100%;"></a></td>
</tr>
<tr>
<th align="right">After</td>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231314010-c2388d2c-8ac6-44f4-b015-f5816493ddde.png"><img src="https://user-images.githubusercontent.com/127789813/231314010-c2388d2c-8ac6-44f4-b015-f5816493ddde.png" alt="image" style="max-width: 100%;"></a></td>
</tr>
</table>

The third commit enables tooltips on the menus for execute forwards/execute backwards. `MakeExecuteAction` already set tooltips, but Qt does not show them unless [`toolTipsVisible`](https://doc.qt.io/qt-5/qmenu.html#toolTipsVisible-prop) is true.

The first commit actually is cleanup that does not change behavior at all. It removes two cases where [`popupMode`](https://doc.qt.io/qt-5/qtoolbutton.html#ToolButtonPopupMode-enum) is set despite the corresponding `QToolButton` not having a menu. I checked all uses of `popupMode` throughout renderdoc while working on the second commit and found that `MenuButtonPopup` is otherwise only used for export and save buttons where there is a default action, but did find these cases where `popupMode` appears to have been set by accident.

w-pearson committed:
Remove specification of QToolButton popupMode when there is no menu

Specifically, the open button in the Python shell window and the blue
channel (only) in the texture viewer had values specified even though
they do not have an actual menu associated with the buttons. For the
texture viewer, this was explicitly set to the default value (delayed
popup); for the Python shell window the value is custom. The popup
mode is only used "for tool buttons that has a menu set or contains
a list of actions".

These were added in early commits for each feature, probably by
accident: 0837e66939b264fb9995624664ea2df1a5c40ad1 for the texture
viewer and 2a4596e06a78d6a3d335469afccda7bfeef86665 for the Python
shell.

w-pearson committed:
Use instant popup for shader execute forwards/backwards buttons

There is no action associated with clicking the button itself, only for
the menu associated with it. Using QToolButton::MenuButtonPopup means
that the majority of the button is for the button's own action, and
there is a small arrow next to the button that shows the menu. Since
there's no action, this means that the small menu arrow needs to be
clicked each time instead of the big button, which is annoying.
QToolButton::InstantPopup instead always shows the menu (though it
doesn't have the arrow indicating a menu exists, although this is not
particularly important).

InstantPopup is also already used for the "Insert" button used when
editing a custom shader for the texture viewer. MenuButtonPopup is used
in many other places throughout RenderDoc, but they are all save or
export buttons where there is a meaningful default action.

w-pearson committed:
Enable tooltips for shader execute forwards/backwards menus

MakeExecuteAction already sets tooltips for each QAction it creates,
but these are not used in a QMenu unless the toolTipsVisible property
is set to true.

baldurk reviewed (changes requested):
This looks good overall - I don't remember exactly what the intention was with the menu buttons, the primary reason for it was because I was adding new stepping modes and the number of individual buttons was too confusing. But I can see that requiring the arrow could be annoying for some people.

That said I disagree with this:
> (though it doesn't have the arrow indicating a menu exists, although this is not particularly important).

I do think it's important to have a UI signal for the user that there is actually a dropdown. Otherwise they may never click on it because they didn't expect it to display a menu instead of immediately taking an action.

Instead I think the styling should be preserved to keep the arrow, but instead make the button display the menu rather than doing anything. That way there's a UI indication that the menu exists but you don't have to click on the arrow for it to work.

w-pearson commented:
The "..." somewhat conveys that a menu would appear, but I'll see if there's a way to reintroduce the arrow without it being drawn as two separate buttons. If I can't find a good way then I'll change it so that the button opens the menu as well.

(The custom shader editor already behaves like this: its insert button has a "..." but no arrow, though "Insert..." has no clear meaning on its own while "Execute forwards..." at least seems like it does. The other uses of `InstantPopup` are the "replay context" menu at the bottom of the screen and the various extension buttons.)

There is a minor downside to using `MenuButtonPopup` but showing the menu on click. At least with this implementation:

```C++
QObject::connect(ui->execForwards, &QToolButton::clicked,
                 [this]() { ui->execForwards->showMenu(); });
```

you have to press and release on the button for the menu to appear, while on the arrow or with `InstantPopup` the menu pops up when the mouse is pressed and if you move the cursor over a menu item and then release, it acts as if you clicked the menu item, although it _also_ works properly with press and release. (Perhaps there's a different signal that can be used for this behavior, though.) This behavior isn't Qt-specific; Visual Studio, Firefox (on the back button), and Notepad++ (on the menu bar) all behave this way. I think this behavior is more noticeable when using a mousepad on a laptop (at least that's where I find myself using it) but I'm not entirely sure of that.

**EDIT**: Using the `QToolButton::pressed` signal gives that same behavior, so that's a workable solution (with the only remaining downside being the line that appears when hovering over the button making it seem that the arrow and the button have different behavior).

w-pearson committed:
Move subControlRect for CC_ToolButton to RDTweakedNativeStyle

This means the full rect won't be returned for SC_ToolButtonMenu if it
is used in a context where the default style doesn't expect the menu.

This is a separate commit to isolate future changes to this function.

w-pearson committed:
Draw arrows next to all tool buttons with menus

Before, only MenuButtonPopup had an arrow. This makes it more obvious
as to whether the button will immediately do something when clicked or
if it will open a menu to choose a specific action.

w-pearson commented:
OK, I figured out how to change the style to show the arrow for any tool button with a menu. This affects multiple parts of RenderDoc's UI, so here are some examples:

<table role="presentation">
<tr><th></th><th>RDStyle</th><th>Native</th></tr>
<tr>
<th align="right">Before</th>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231618008-845fdd10-36b1-4b02-b61d-a8595b71adc5.png"><img src="https://user-images.githubusercontent.com/127789813/231618008-845fdd10-36b1-4b02-b61d-a8595b71adc5.png" /></a></td>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231618019-3259c611-a40e-4cf7-a4a3-ff53d9d4a12f.png"><img src="https://user-images.githubusercontent.com/127789813/231618019-3259c611-a40e-4cf7-a4a3-ff53d9d4a12f.png" /></a></td>
</tr><tr>
<th>After</th>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231618014-43c145b5-173f-46fd-b3c6-c61d874de859.png"><img src="https://user-images.githubusercontent.com/127789813/231618014-43c145b5-173f-46fd-b3c6-c61d874de859.png" /></a></td>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231618026-fc4052e9-46b5-4336-8d50-9dabadd3ff0b.png"><img src="https://user-images.githubusercontent.com/127789813/231618026-fc4052e9-46b5-4336-8d50-9dabadd3ff0b.png" /></a></td>
</tr>
</table>

<table role="presentation">
<tr><th></th><th>RDStyle</th><th>Native</th></tr>
<tr>
<th align="right">Before</th>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231618040-505c748a-4909-4ab3-9b80-dbe19c7374a6.png"><img src="https://user-images.githubusercontent.com/127789813/231618040-505c748a-4909-4ab3-9b80-dbe19c7374a6.png" /></a></td>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231618053-430c5528-2917-40e4-9ee6-6c3363a80ba6.png"><img src="https://user-images.githubusercontent.com/127789813/231618053-430c5528-2917-40e4-9ee6-6c3363a80ba6.png" /></a></td>
</tr><tr>
<th>After</th>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231618048-8f03054b-35c0-48a8-a85e-6c4d853879bf.png"><img src="https://user-images.githubusercontent.com/127789813/231618048-8f03054b-35c0-48a8-a85e-6c4d853879bf.png" /></a></td>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231618058-8e9becfa-a6f9-481f-b289-8e492b62cada.png"><img src="https://user-images.githubusercontent.com/127789813/231618058-8e9becfa-a6f9-481f-b289-8e492b62cada.png" /></a></td>
</tr>
</table>

<table role="presentation">
<tr><th></th><th>RDStyle</th><th>Native</th></tr>
<tr>
<th align="right">Before</th>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231618065-89978d3c-b52d-4458-98d3-0b72a0ab4ebe.png"><img src="https://user-images.githubusercontent.com/127789813/231618065-89978d3c-b52d-4458-98d3-0b72a0ab4ebe.png" /></a></td>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231618076-70364beb-9d0e-4797-9bb9-550198748a19.png"><img src="https://user-images.githubusercontent.com/127789813/231618076-70364beb-9d0e-4797-9bb9-550198748a19.png" /></a></td>
</tr><tr>
<th>After</th>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231618070-96f39b17-f392-4b6f-976b-6ec141b4c093.png"><img src="https://user-images.githubusercontent.com/127789813/231618070-96f39b17-f392-4b6f-976b-6ec141b4c093.png" /></a></td>
<td align="left"><a href="https://user-images.githubusercontent.com/127789813/231618083-dad42e70-7536-42ac-bbb5-a7b7688cd9bb.png"><img src="https://user-images.githubusercontent.com/127789813/231618083-dad42e70-7536-42ac-bbb5-a7b7688cd9bb.png" /></a></td>
</tr>
</table>

This was a bit complicated but not *too* hard since there already were some style customizations for the tool buttons.

baldurk commented:
Thanks, that looks good to me and yes I think this is better for the other cases where the button unconditionally pops up a menu too.
